clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -244,6 +267,50 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void getPID(long remainingTime) {
+        // Send an InitPIDRequest, wait for the response, retrieve the PID from the response if success. Return it.
+        // Otherwise throw exception.
+        // Pick a random broker to send the initPIDRequest to.
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
+            client.poll(100, time.milliseconds());
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+        if (remainingTime <= 0) {
+            return;
+        }
+
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleInitPidResponse(response);
```

Comment:
It is just for convenience for now. The client used on the consumer has synchronous calls which use callbacks internally. We could unify the two clients in the future, and do away with this call back then.